datasource/src/main/scala/quasar/blobstore/azure/AzureGetService.scala

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2014–2018 SlamData Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.blobstore.azure
+
+import slamdata.Predef._
+import quasar.blobstore.azure.requests.DownloadArgs
+import quasar.blobstore.services.GetService
+
+import cats.data.Kleisli
+import cats.effect.ConcurrentEffect
+import cats.syntax.applicative._
+import com.microsoft.azure.storage.blob._
+import com.microsoft.rest.v2.Context
+import fs2.Stream
+
+object AzureGetService {
+
+  def apply[F[_]: ConcurrentEffect, P](
+      toBlobUrl: Kleisli[F, P, BlobURL],
+      mkArgs: BlobURL => DownloadArgs,
+      reliableDownloadOptions: ReliableDownloadOptions,
+      maxQueueSize: MaxQueueSize,
+      errorHandler: P => Throwable => Stream[F, Byte])
+      : GetService[F, P] = {
+
+    val getSvc = toBlobUrl map
+      mkArgs andThen
+      requests.downloadRequestK andThen
+      handlers.toByteStreamK(reliableDownloadOptions, maxQueueSize)
+
+    Kleisli[F, P, Stream[F, Byte]] { p =>
+      Stream.force(getSvc(p)).handleErrorWith(errorHandler(p)).pure[F]
+    }
```

Comment:
@wemrysi If I understood you correctly, you intended this to be `GetService[F, BlobURL]`? I played a bit with that but kept it `GetService[F, P]` in the end. Mainly because of the `errorHandler` which I imagine one typically doesn't want to express in terms of `BlobURL`.

Any way, this code here is meant to be at a higher level than the building blocks in `requests` and `handlers`. The building blocks are meant to be functionally complete, but the code here should just be sufficient for most use cases - in case it's not sufficient then it's always possible to cherry pick the required building blocks and create your own pipeline (same holds for the other `Azure<x>Service`s).

Reply:
The intent of these `Azure*Service`s is to provide converters that are generic in `P` and/or `R`. These functions are meant as templates / common scenarios, like typically one would want a `PropsService[F, MyPath, MyRes]` and `PropsService.apply/mk` would be applicable.

E.g. in our case, I do like
```
        AzurePropsService.mk[F, BlobPath, Boolean](
          blobPathToBlobURLK,
          Kleisli(_ => true.pure[F]),
          _.recover { case _: StorageException => false }),
```

It's possible to construct pipelines like this at call site, but I think there's value in providing some templates (e.g. before I discovered `mapF` I had clumsier ways to express the same pipelines), in particular when they seem to be generally applicable (like for `AzurePropsService`). Another thing is that I'm looking for the minimal amount of information (think cardinality) that needs to be provided in order to be able to construct a generic `Service`.

I have improved `AzureGetService` but can imagine the errorHandler part can be improved further (decoupled somehow maybe).

Maybe discuss further in chat?